cmd/kops/toolbox_instance_selector.go

```diff
@@ -0,0 +1,487 @@
+	}
+
+	cluster, err := clientset.GetCluster(ctx, clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if cluster == nil {
+		return nil, nil, nil, fmt.Errorf("cluster %q not found", clusterName)
+	}
+
+	channel, err := cloudup.ChannelForCluster(cluster)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if len(cluster.Spec.Subnets) == 0 {
+		return nil, nil, nil, fmt.Errorf("configuration must include Subnets")
+	}
+
+	return clientset, cluster, channel, nil
+}
+
+func getFilters(commandline *cli.CommandLineInterface, region string) selector.Filters {
+	flags := commandline.Flags
+	return selector.Filters{
+		VCpusRange:             commandline.IntRangeMe(flags[vcpus]),
+		MemoryRange:            commandline.IntRangeMe(flags[memory]),
+		VCpusToMemoryRatio:     commandline.Float64Me(flags[vcpusToMemoryRatio]),
+		CPUArchitecture:        commandline.StringMe(flags[cpuArchitecture]),
+		GpusRange:              commandline.IntRangeMe(flags[gpus]),
+		GpuMemoryRange:         commandline.IntRangeMe(flags[gpuMemoryTotal]),
+		PlacementGroupStrategy: commandline.StringMe(flags[placementGroupStrategy]),
+		UsageClass:             commandline.StringMe(flags[usageClass]),
+		EnaSupport:             commandline.BoolMe(flags[enaSupport]),
+		Burstable:              commandline.BoolMe(flags[burstSupport]),
+		Region:                 commandline.StringMe(region),
+		AvailabilityZones:      commandline.StringSliceMe(flags[availabilityZones]),
+		MaxResults:             commandline.IntMe(flags[maxResults]),
+		NetworkInterfaces:      commandline.IntRangeMe(flags[networkInterfaces]),
+		NetworkPerformance:     commandline.IntRangeMe(flags[networkPerformance]),
+		AllowList:              commandline.RegexMe(flags[allowList]),
+		DenyList:               commandline.RegexMe(flags[denyList]),
+		InstanceTypeBase:       commandline.StringMe(flags[instanceTypeBase]),
+		Flexible:               commandline.BoolMe(flags[flexible]),
+	}
+}
+
+func getInstanceSelectorOpts(commandline *cli.CommandLineInterface) InstanceSelectorOptions {
+	opts := InstanceSelectorOptions{}
+	flags := commandline.Flags
+	opts.NodeCountMin = int32(*commandline.IntMe(flags[nodeCountMin]))
+	opts.NodeCountMax = int32(*commandline.IntMe(flags[nodeCountMax]))
+	opts.InstanceGroupName = *commandline.StringMe(flags[igName])
+	opts.Output = *commandline.StringMe(flags[output])
+	opts.DryRun = *commandline.BoolMe(flags[dryRun])
+	opts.ClusterAutoscaler = *commandline.BoolMe(flags[clusterAutoscaler])
+	if flags[nodeVolumeSize] != nil {
+		volumeSize := int32(*commandline.IntMe(flags[nodeVolumeSize]))
+		opts.NodeVolumeSize = &volumeSize
+	}
+	if flags[nodeSecurityGroups] != nil {
+		opts.NodeSecurityGroups = *commandline.StringSliceMe(flags[nodeSecurityGroups])
+	}
+	if flags[instanceGroupCount] != nil {
+		opts.InstanceGroupCount = *commandline.IntMe(flags[instanceGroupCount])
+	}
+	return opts
+}
+
+func getClusterZones(subnets []kops.ClusterSubnetSpec) ([]string, error) {
```

Comment:
I think if the availabilityZones flag is provided then the instance group should only span those zones rather than all zones of the cluster.